### PR TITLE
Git Zones and Knot DNS Server

### PR DESCRIPTION
**Pull request by nils-wisiol**
This PR changes the replication source from AXFR and notification and catalog based to zonesfiles from a git repository. This is the PR corresponding to [desec-io/desec-stack#454](https://github.com/desec-io/desec-stack/pull/454).

It also changes the auth nameserver from pdns and dnsdist to knot DNS.

- [x] discuss if dnsdist should remain at the front
- [x] find a replacement for pdns metronome metics
- [ ] TODOs from the code

### CODE REVIEW

**Comment by peterthomassen** [comment]
I think it would be a good idea to have dnsdist in front, for abuse mitigation, DoH (if we want) etc.pp. Also, metrics currently are in fact from dnsdist (pdns didn't support them yet, but does now). If we keep dnsdist, we get the knot metrics for free.

File: replicator/replicator/git.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            self._git_do('remote', 'add', 'origin', origin)
+
+    def get_head(self):
+        return self.get_commit('HEAD')
+
+    def get_commit(self, rev):
+        try:
+            commit_msg = self._git_do('show', rev, '--format=%s', '-s')[:-1]
+            commit_hash = self._git_do('show', rev, '--format=%H', '-s')[:-1]
+            return commit_msg, commit_hash
```
Comment (peterthomassen) [inline replicator/replicator/git.py:94]: inconsistent with what's in desec-stack (no big deal, but may lead to confusion later)

**Comment by peterthomassen** [comment]
Also, IMO the pdns/lmdb removal and the knot addition should be two different commits. (Also, ideally, the docker-compose application would support switching between both backends ... but let's not waste too much time on that.)

**Comment by nils-wisiol** [comment]
To recover the git zones repo after the server already deleted history (e.g. by not pulling for more than a couple of days), use

    git fetch --update-shallow
    git reset --hard origin/main

Of course, there will be no diff information, so that needs to be done by the recovery method of the replicator client

**Comment by nils-wisiol** [comment]
Make sure the DNS server survives `docker-compose down && docker-compose up -d`

**Comment by nils-wisiol** [comment]
In case `git pull` fails because history was truncated too far at the server, git will say

```
warning: no common commits
remote: Enumerating objects: 51756, done.
remote: Counting objects: 100% (51756/51756), done.
remote: Compressing objects: 100% (31555/31555), done.
remote: Total 51756 (delta 20254), reused 51661 (delta 20201), pack-reused 0
Receiving objects: 100% (51756/51756), 71.15 MiB | 3.45 MiB/s, done.
Resolving deltas: 100% (20254/20254), done.
warning: reject refs/remotes/origin/main because shallow roots are not allowed to be updated
Your configuration specifies to merge with the ref 'refs/heads/main'
from the remote, but no such ref was fetched.
```

**Comment by nils-wisiol** [comment]
For debugging, the replicator could have a small log database where for each git revision, publication timestamp is stored.

**Comment by peterthomassen** [comment]
> In case git cannot update because history was truncated too far at the server, git will say

Could fall-back on a fresh clone then.

**Comment by nils-wisiol** [comment]
1. Sometimes, the replicators `git clone` fails because there is a `timers` directly inside `/zones/`, which stops git from cloning with error message "fatal: destination path '.' already exists and is not an empty directory." The directory was [probably created by knot](https://www.knot-dns.cz/docs/2.9/html/reference.html#timer-db). We could move it using the knot config or delete it prior to the fresh clone.
2. When the SSH connection to desec stack fails, there is no information in the log how to fix the situation. desec-ns should display the information required to add the desec-ns public key to the authorized keys at desec-stack. There are two possible reasons:
    1. The remote host id changed and is not accepted, in which case the output of `docker-compose exec gitzones auth id` from desec-stack is helpful
    1. The key pair of desec-ns is not authorized to connect, in which case the public key should be displayed in the logs and can be authorized with `docker-compose exec gitzones auth add ....` at desec-stack.

**Comment by nils-wisiol** [comment]
desec-ns should clean the repository every once in a while to avoid exhausting the storage space

**Comment by nils-wisiol** [comment]
desec-ns should observe if commands to knot fail and act upon it, e.g. in cases such as

```
May 13 16:47:17 ubuntu desec-ns/replicator[1294]: INFO:replicator.nameserver:>>> KNOTC >>> zone-reload 'XXXXX.dedyn.io'
May 13 16:47:17 ubuntu desec-ns/ns[1294]: 2021-05-13T16:47:17+0000 info: [XXXXX.dedyn.io.] control, received command 'zone-reload'
May 13 16:47:17 ubuntu desec-ns/ns[1294]: 2021-05-13T16:47:17+0000 error: [XXXXX.dedyn.io.] control, error (no such zone found)
```

**Comment by peterthomassen** [comment]
Make sure to support HTTPS/SVCB records (Knot has [tentative plans for 3.1](https://gitlab.nic.cz/knot/knot-dns/-/issues/654))

**Comment by nils-wisiol** [comment]
desec-ns needs to truncate the local history from time to time to reduce and limit storage consumption. The history that needs to be kept can be very small.

History can be truncated using `Repository.remove_history`:

```
>>> from replicator.git import Repository
>>> from datetime import datetime, timedelta
>>> before = datetime.now() - timedelta(days=2)
>>> zones = Repository('/zones')
>>> zones.remove_history(before=before)
```

**Comment by nils-wisiol** [comment]
Make sure that the pid file of knot is deleted prior to start up. Otherwise start up may be inhibited by a stale pid file after hard shutdowns.

**Comment by nils-wisiol** [comment]
knot can take its time processing our input. This can lead to 

`subprocess.TimeoutExpired: Command '['/usr/sbin/knotc', '-s', '/rundir/knot/knot.sock']' timed out after 30 seconds`

**Comment by peterthomassen** [comment]
Turned out insufficiently stable on the primary, abandoning. Thanks, @nils-wisiol, for the hard work! :clap:

**Pull request closed without merging**